### PR TITLE
[RFC 65] Add custom namespace columns to CNA

### PR DESCRIPTION
**Pull request by BasLee**
Add custom namespace columns to CNA events

image

## Changes
- Add `ANNOTATION_JSON` column to `sample_cna_event table`
- Add [CustomNamespaceParser.java](https://github.com/BasLee/cbioportal/compare/custom-cna-namespace...BasLee:cbioportal:custom-cna-namespace-annotation-json?expand=1#diff-8133c497150e14bfb30c8c96682d8ef8cdb2f37e32784592706224c0a8179283) that imports custom columns of both mutations and CNA's
- Expose the custom columns as `namespaceColumns` in the `DiscreteCopyNumberController` (see `DiscreteCopyNumberDataMixin`)

## Tests
- Cna importer imports custom namespace columns, see: `TestImportCnaDiscreteLongData`
- API exposes custom namespace columns, see: `DiscreteCopyNumberServiceImplTest` and `DiscreteCopyNumberControllerTest`

# Checks
- [ ] When importing es0, custom namespace columns should be added to the response of [./api/molecular-profiles/study_es_0_gistic/discrete-copy-number](http://0.0.0.0:8081/api/molecular-profiles/study_es_0_gistic/discrete-copy-number?sampleListId=study_es_0_cna&discreteCopyNumberEventType=HOMDEL_AND_AMP) :
 ```
response.filter(e => e.entrezGeneId === 116983 && e.sampleId === 'TCGA-A2-A04U-01').map(e => e.namespaceColumns)
{
    "MyNamespace": {
        "column1": "value1b",
        "column2": "value2b"
    },
    "MyNamespace2": {
        "blarp": "bluirp2"
    }
}
```

### CODE REVIEW

File: core/src/main/java/org/mskcc/cbio/portal/util/CnaUtil.java [hunk 1]
```diff
@@ -70,9 +80,18 @@ public CnaEvent createEvent(GeneticProfile geneticProfile, int sampleId, String[
...
         cna.setDriverTiersFilter(TabDelimitedFileUtil.getPartString(getColumnIndex(CnaUtil.CBP_DRIVER_TIERS), parts));
         cna.setDriverTiersFilterAnnotation(TabDelimitedFileUtil.getPartString(getColumnIndex(CnaUtil.CBP_DRIVER_TIERS_ANNOTATION), parts));
+        cna.setAnnotationJson(
+            convertMapToJsonString(
+                this.namespaceColumnParser.parseCustomNamespaces(parts)
+            )
+        );
         return cna;
     }
-
```
Comment (Luke-Sikina) [inline core/src/main/java/org/mskcc/cbio/portal/util/CnaUtil.java:75]:
```java
    public int getColumnIndex(String colName) {
        Integer index = this.columnIndexMap.get(colName.toLowerCase());
        if (index == null) {
            index = -1;
        }
        return index;
    }
```

This entire function is annoying me. Can we at least make it private? It's only used internally. Bonus points if you switch the `get` for `getOrDefault`

File: model/src/main/java/org/cbioportal/model/DiscreteCopyNumberData.java [hunk 2]
```diff
@@ -1,18 +1,33 @@
...
 import javax.validation.constraints.NotNull;
 
 public class DiscreteCopyNumberData extends Alteration implements Serializable {
-    
     @NotNull
     private Integer alteration;
-    
+
+    @JsonRawValue
+    @ApiModelProperty(dataType = "java.util.Map")
```
Comment (Luke-Sikina) [inline model/src/main/java/org/cbioportal/model/DiscreteCopyNumberData.java:14]: I had no idea this was a feature. Neat!

File: model/src/main/java/org/cbioportal/model/MolecularProfile.java [hunk 3]
```diff
@@ -36,6 +36,22 @@ public enum ImportType {
...
+        public static CnaDatatype findByName(String name) {
+            CnaDatatype result = null;
+            for (CnaDatatype value : values()) {
+                if (value.name().equalsIgnoreCase(name)) {
+                    result = value;
+                    break;
+                }
+            }
+            return result;
+        }
```
Comment (Luke-Sikina) [inline model/src/main/java/org/cbioportal/model/MolecularProfile.java:52]: Guava has a nice method for this. We don't import guava here, but we do in the service layer. You could replace that method with this code, and delete the logic here:

```java
    private boolean isDiscreteCnaDataType(MolecularProfile molecularProfile) {
        return Enums
            .getIfPresent(MolecularProfile.CnaDatatype.class, molecularProfile.getDatatype().toUpperCase())
            .isPresent();
    }
```

Comment (Luke-Sikina) [reply model/src/main/java/org/cbioportal/model/MolecularProfile.java:52]: Oh, the diff changed, and now this isn't used. Wrong PR?

**Review by Luke-Sikina — APPROVED** [review]
Very small changes. Nice job!

File: core/src/main/java/org/mskcc/cbio/portal/util/CnaUtil.java (hunk 1 above)
Comment (BasLee) [reply core/src/main/java/org/mskcc/cbio/portal/util/CnaUtil.java:75]: Fixed in [7bccc41](https://github.com/cBioPortal/cbioportal/compare/9ce311d9b4a4cabb7012450c85e285ba553fa5af..7bccc41e1e6dd57d97f8e258889a730f8fb019a7)

File: model/src/main/java/org/cbioportal/model/DiscreteCopyNumberData.java (hunk 2 above)
Comment (BasLee) [reply model/src/main/java/org/cbioportal/model/DiscreteCopyNumberData.java:14]: Stolen from mutation.annotationJson :)

File: core/src/main/java/org/mskcc/cbio/portal/scripts/NamespaceColumnParser.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            return Float.parseFloat(stringValue);
+        } else if (isDouble(stringValue)) {
+            return Double.parseDouble(stringValue);
+        }
+        return stringValue;
+    }
+
+    private String convertMapToJsonString(Map<String, Map<String, Object>> map) throws IOException {
+        return this.mapper.writeValueAsString(map);
+    }
```
[the remark this answers is not in the document]
Comment (BasLee) [reply core/src/main/java/org/mskcc/cbio/portal/scripts/NamespaceColumnParser.java]: Fixed in [7bccc41](https://github.com/cBioPortal/cbioportal/compare/9ce311d9b4a4cabb7012450c85e285ba553fa5af..7bccc41e1e6dd57d97f8e258889a730f8fb019a7)

File: core/src/main/java/org/mskcc/cbio/portal/scripts/ImportExtendedMutationData.java [hunk 5]
```diff
@@ -421,7 +421,7 @@ public void importData() throws IOException, DaoException {
                         ascn = new AlleleSpecificCopyNumber(ascnData);
                     }
                     if (record.getNamespacesMap() != null && !record.getNamespacesMap().isEmpty()) {
-                        mutation.setAnnotationJson(convertMapToJsonString(record.getNamespacesMap()));
+                        mutation.setAnnotationJson(this.convertMapToJsonString(record.getNamespacesMap()));
```
[the remark this answers is not in the document]
Comment (BasLee) [reply core/src/main/java/org/mskcc/cbio/portal/scripts/ImportExtendedMutationData.java]: Fixed in [7bccc41](https://github.com/cBioPortal/cbioportal/compare/9ce311d9b4a4cabb7012450c85e285ba553fa5af..7bccc41e1e6dd57d97f8e258889a730f8fb019a7)

File: core/src/main/java/org/mskcc/cbio/portal/scripts/NamespaceColumnParser.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
+package org.mskcc.cbio.portal.scripts;
+
```
[the remark this answers is not in the document]
Comment (BasLee) [reply core/src/main/java/org/mskcc/cbio/portal/scripts/NamespaceColumnParser.java]: Fixed in [7bccc41](https://github.com/cBioPortal/cbioportal/pull/9878/commits/7bccc41e1e6dd57d97f8e258889a730f8fb019a7)

File: model/src/main/java/org/cbioportal/model/MolecularProfile.java (hunk 3 above)
Comment (BasLee) [reply model/src/main/java/org/cbioportal/model/MolecularProfile.java:52]: Related to these changes: https://github.com/cBioPortal/cbioportal/pull/9847#issuecomment-1318605217

**Review by dippindots — APPROVED** [review]

**Pull request merged**